[Pytorch Edge] get_model_bytecode int -> uint

JacobSzwejbka opened the pull request:
Summary: change int to uint to be the same type as the runtimes bytecode. Only affects c++ since python doesn't have uints iirc. Also changed the behavior of the functions from returning -1 and a warning to just throw an exception. Wasnt sure what the proper behavior here would be (returning UINT_MAX seemed gross) so feedback is appreciated.

Test Plan: ci

Differential Revision: D29914072

raziel commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -58,38 +58,38 @@ std::vector<IValue> get_bytecode_ivalues(PyTorchStreamReader& reader) {
     const std::vector<IValue>& bytecode_ivalues) {
   if (!bytecode_ivalues.empty() && bytecode_ivalues[0].isInt()) {
-    int64_t model_version = bytecode_ivalues[0].toInt();
+    uint64_t model_version = static_cast<uint64_t>(bytecode_ivalues[0].toInt());
```
I'm assuming at this point bytecode_ivalues[0] should have been checked against a non negative value?

JacobSzwejbka replied:
Yeah that makes sense, I can add that validation.

raziel reviewed (approved):
Thanks!
A couple of comments on general validation of bytecode.
Probably good for @iseeyuan to take a look as well.

JacobSzwejbka commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -58,38 +58,38 @@ std::vector<IValue> get_bytecode_ivalues(PyTorchStreamReader& reader) {
-    TORCH_WARN(
-        "The input model might not be generated from _save_for_mobile()");
-    return -1;
+    TORCH_CHECK(
```
cc @xcheng16

cccclai replied:
The feedback I received from FL team is to exit gracefully, instead of throwing error. There will be some validation process running in server and client, and gracefully shutdown seems reasonable to me. That's the reason I've been using `TORCH_WARN`. cc @fbhuba

JacobSzwejbka commented:
We considered making this optional<uint> or returning UINTMAX to preserve the current TORCH_WARN behavior, but ultimately decided that if theres no bytecode version in the model file, its an ill formed model file and we should probably just crash.

cccclai commented:
> We considered making this optional or returning UINTMAX to preserve the current TORCH_WARN behavior, but ultimately decided that if theres no bytecode version in the model file, its an ill formed model file and we should probably just crash.

This api (and other api under `model_compatibility.cpp`) is not for validation purpose. The assumption is that, the input model is supposed to be a valid model file. It will just perform some minimum checks. This api has been used in model delivery server and crash is probably not a good experience for clients.

dhruvbird commented:
I feel like consumers of PyTorch should be ready to handle `c10::Error` at the very least when calling into a PyTorch API. Based on that, it may be okay to throw since it's a catchable exception (i.e. not a process abort).

cccclai commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -58,38 +58,39 @@ std::vector<IValue> get_bytecode_ivalues(PyTorchStreamReader& reader) {
   if (!bytecode_ivalues.empty() && bytecode_ivalues[0].isInt()) {
     int64_t model_version = bytecode_ivalues[0].toInt();
-    return model_version;
+    TORCH_CHECK(model_version > 0, "Model bytecode version < 0");
```
`TORCH_CHECK(model_version > 0, "Expect model bytecode version > 0, but got ",  model_version);`

cccclai commented:
> I feel like consumers of PyTorch should be ready to handle `c10::Error` at the very least when calling into a PyTorch API. Based on that, it may be okay to throw since it's a catchable exception (i.e. not a process abort).

It doesn't sound very sustainable to let user rely on try catch to avoid crash. It means server side needs to use try catch whenever they want to check a bytecode version. I'd prefer previous logic. If the new flow is strongly preferred, we can go for it, but we need to make sure all existing users are aware of this change and change their code accordingly if it's needed.

JacobSzwejbka commented:
> > I feel like consumers of PyTorch should be ready to handle `c10::Error` at the very least when calling into a PyTorch API. Based on that, it may be okay to throw since it's a catchable exception (i.e. not a process abort).
> 
> It doesn't sound very sustainable to let user rely on try catch to avoid crash. It means server side needs to use try catch whenever they want to check a bytecode version. I'd prefer previous logic. If the new flow is strongly preferred, we can go for it, but we need to make sure all existing users are aware of this change and change their code accordingly if it's needed.

I'm not sure I understand when it would be reasonable to pass in something that isnt a model file to this api. If your checking to see what the bytecode version is of this model, and you pass in an ill formatted model (or something thats not a model) that sounds like a bug to me, and making the bug loud (an exception) seems preferable to silently returning a -1 equivalent.

cccclai commented:
> > > I feel like consumers of PyTorch should be ready to handle `c10::Error` at the very least when calling into a PyTorch API. Based on that, it may be okay to throw since it's a catchable exception (i.e. not a process abort).
> > 
> > 
> > It doesn't sound very sustainable to let user rely on try catch to avoid crash. It means server side needs to use try catch whenever they want to check a bytecode version. I'd prefer previous logic. If the new flow is strongly preferred, we can go for it, but we need to make sure all existing users are aware of this change and change their code accordingly if it's needed.
> 
> I'm not sure I understand when it would be reasonable to pass in something that isnt a model file to this api. If your checking to see what the bytecode version is of this model, and you pass in an ill formatted model (or something thats not a model) that sounds like a bug to me, and making the bug loud (an exception) seems preferable to silently returning a -1 equivalent.
> 
> If we wanted to go down the side of returning some special value that means "not a model file" like how some apis will return null ptr or -1, then I'm not sure what a good return value is. Optional, 0, or UINT_MAX seem kinda gross and unintuitive to me.

Ideally a problematic model shouldn't be passed to this api. However server is a complicated system and it's hard to predict all cases, consider all threading and network etc. We just don't want to bring the system down because failing to get bytecode version from a malformed model.

JacobSzwejbka commented:
Ok. If we really want to maintain the soft failure instead of a crash, then I think optional is the best way to do it for uints. I dont think the confusion about "bytecode version" not really being an optional field is a huge concern, since these apis are mostly supposed to be obscured from the user, and they are supposed to use the black box api in the follow up diff.

JacobSzwejbka commented:
Verified that the exceptions thrown by TORCH_CHECK can be caught and dont just trigger sigabort @fbhuba @raziel. Was going to add a unit test that triggered and caught them, but I was finding it pretty cumbersome to try and manually trash bytecode.pkl inside the archive.

JacobSzwejbka committed:
[Pytorch Edge] get_model_bytecode int -> uint (#62201)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/62201

change int to uint to be the same type as the runtimes bytecode. Only affects c++ since python doesn't have uints iirc. Also changed the behavior of the functions from returning -1 and a warning to just throw an exception. Wasnt sure what the proper behavior here would be (returning UINT_MAX seemed gross) so feedback is appreciated.

Test Plan: ci

Reviewed By: raziel

Differential Revision: D29914072

fbshipit-source-id: e9e6d0ea4823c140a957edec362e4da9a3975ff2
